tangelo/toolboxes/circuits/qsp.py

```diff
@@ -0,0 +1,272 @@
+        qubcirc = Circuit([Gate("CH", anc, control=c_list), Gate("CZ", anc, control=c_list)])
+    else:
+        qubcirc = Circuit([Gate("CH", anc, control=c_list)])
+
+    circ_angles = []
+    circ_angles = [angles[0]+np.pi/4]
+    for ang in angles[1:-1]:
+        circ_angles += [ang + np.pi/2]
+    circ_angles += [angles[-1]+np.pi/4]
+
+    zcnot = zero_controlled_cnot(m_qs, [anc], [])
+
+    qubcirc += zcnot + Circuit([Gate("CRZ", target=anc, parameter=2*circ_angles[-1], control=c_list)]) + zcnot + cua
+    for j, ang in enumerate(circ_angles[-2:0:-1]):
+        qubcirc += Circuit([Gate("CZ", anc, control=c_list)]) + zcnot + Circuit([Gate("CRZ", target=anc, parameter=2*ang, control=c_list)]) + zcnot + cua
+    qubcirc += Circuit([Gate("CZ", anc, control=c_list)]) + zcnot + Circuit([Gate("CRZ", target=anc, parameter=2*circ_angles[0], control=c_list)]) + zcnot
+
+    qubcirc += Circuit([Gate("CH", anc, control=c_list)])
+
+    return qubcirc
+
+
+def get_qsp_hamiltonian_simulation_qubit_list(qu_op: QubitOperator) -> List[int]:
+    "Returns the list of qubits used for the QSP Hamiltonian simulation algorithm"
+    qu_op_qs, uprep_qs, _ = get_lcu_qubit_op_info(qu_op)
+    return qu_op_qs + uprep_qs + list(range(uprep_qs[-1]+1, uprep_qs[-1]+4))
+
+
+def get_qsp_hamiltonian_simulation_circuit(qu_op: QubitOperator, tau: float, eps: float = 1.e-4, control: Union[int, List[int]] = None, n_attempts: int = 10,
+                                           method: str = 'laurent', folder: str = None) -> Circuit:
+    """Returns Quantum Signal Processing (QSP) Hamiltonian simulation circuit for a given QubitOperator for time tau.
+
+    The circuits are derived from https://arxiv.org/abs/2002.11649.
+    The list of qubits used for the circuit can be found by qubit_list = get_qs_hamiltonian_simulation_qubit_list(qu_op)
+    qu_op must have it's spectral range in [-1, 1]. This property is not checked by this function.
+
+    Args:
+        qu_op (QubitOperator): The qubit operator defining the QSP time-evolution circuit.
+        tau (float): The evolution time.
+        eps (float): The allowed error of the time-evolution. Higher accuracy requires longer circuits, and obtaining the phases can be unstable.
+        control (Union[int, List[int]]): The control qubit(s).
+        n_attempts (int): The number of attempts to calculate the phase factors using pyqsp.
+        method (str): "laurent": Use laurent polynomial method of pyqsp, "tf": Use TensorFlow with pyqsp, "QSPPACK": Use QSPPACK to calculate phases
+        folder (str): Folder that contains QSPPACK.
+
+    Returns:
+        Circuit: The QSP Hamiltonian simulation circuit with oblivious amplitude amplification to ensure very high success probability.
+    """
+
+    if control is not None:
+        control_list = copy(control) if isinstance(control, list) else [control]
+    else:
+        control_list = []
+
+    # If tau is negative, flip sign of tau and qu_op.
+    flip_tau_sign = False
```

Comment:
set that variable to `flip_tau_sign = (tau < 0.)`.
Then I think L221-224 can be replaced by a duplicate of L268-270.